builtin/credential/aws-ec2/client.go

```diff
@@ -97,28 +118,41 @@ func (b *backend) clientEC2(s logical.Storage, region string) (*ec2.EC2, error)
 	defer b.configMutex.Unlock()
 
 	// If the client gets created while switching the locks, return it
-	if b.EC2ClientsMap[region] != nil {
-		return b.EC2ClientsMap[region], nil
+	if b.EC2ClientsMap[region][stsRole] != nil {
```

Comment:
The same problem exists here. Generally speaking you should probably fetch/check this once up top and then use that value later (or not, if it doesn't exist).